### PR TITLE
Multi-column sorting

### PR DESCRIPTION
**Pull request by wszymanski**
### Context
Introducing the new PRO plugin.

# UI & UX

- New sorting order arrows were introduced, numbers by the arrows indicates sort sequence. 

screen shot 2018-09-20 at 11 50 13

Note: We know that `Firefox` may show numbers lower than Chrome. It's related to the position of `span` inside header.

- Sort indicator is shown by default.
- Click with `cmd` key (`macOS`) or `ctrl` key (other platforms) allow to sort multiple columns.
- Sorting by click may be disabled by `headerAction` option inside configuration of the plugin (`multiColumnSorting` key).
- The non-contiguous selection on headers is blocked when user click on the sorting arrow.
- Column header which may display indicator (by default, option `indicator` is set to `true`) reserves extra space for purpose of showing it (header is wider). It should also work properly when using `disablePlugin`, `enablePlugin`, `updateSettings` methods.
- `span` DOM element inside column header is `inline-block` from now with `line-height` set to `1.1`.

# API & engine

- We can declare some options for particular columns by [`columns`](https://docs.handsontable.com/pro/Options.html#columns) option, but please keep in mind that it was delivered by workaround (`cellMeta` inheriance isn't full yet).
    - `compareFunctionFactory` - a curry function which return [compare function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#Description). It can be set now in configuration of the plugin (`multiColumnSorting` key). It may be also used in `columns` property, i.e.:
```
columns(visualColumn) {
  if (visualColumn === 1) {
    return {
      multiColumnSorting: {
        compareFunctionFactory(sortOrder, columnMeta) {
          return function(value, nextValue) {
            // Some value comparisons which will return -1, 0 or 1...
          }
        }
      }
    }
  }

  return {};
}
```
-
    - `indicator`. It may be stored in configuration of the plugin (`multiColumnSorting` key).
- Comparators

```
function numericSort(sortOrder, columnMeta) {
  return function (value, nextValue) {
```
- New plubic API methods were added:
	- `getSortConfig`
	- `setSortConfigs`
	- `clearSort`
- The `sort` method it's pure function, so every call of `sort` function set an entirely new sort order. Previous sort configs aren't preserved. Example of calls are shown below:

```
hot.getPlugin('multiColumnSorting').sort({ column: 0, sortOrder: 'asc' });

// sort first two visual column in the defined sequence
hot.getPlugin('multiColumnSorting').sort([{
  column: 1, sortOrder: 'asc'
}, {
  column: 0, sortOrder: 'desc'
}]);
```

- Some validations of config handled by `sort` method were added.
- Hooks `beforeColumnSort` and `afterColumnSort` gets as arguments current sort config, destination sort config and flag if sort is possible (it has been developed similar to the https://github.com/handsontable/handsontable/issues/5068 - passing **all** useful information)
- The `updateSettings` with defined `multiColumnSort` set completly new sort settings (nothing is preserved).

### How has this been tested?

Manual tests of API and work of the `PersistantState` plugin. 

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Related issue(s):
1. #94 
2. https://github.com/handsontable/handsontable/issues/5282

### Checklist:

- [x] My code follows the code style of this project,
- [x] My change requires a change to the documentation.

### CODE REVIEW

**Comment by wszymanski** [comment]
Next changes and improvements compared to the current version of the ColumnSorting plugin:

- The disabling plugin also clears sort indicators
- The `headerAction` option should work with `ManualColumnMove` properly
- Sorting arrows and numeric indicators should make some place for them in the headers.
- Extra warn about enabling `columnSorting` and `multiColumnSorting` plugins.
- Mose moves beetween sorting indicators and header text should be smooth.

Note: Please keep in mind that headers are now much wider.

**Comment by wszymanski** [comment]
Hooks have been deloped similar to the https://github.com/handsontable/handsontable/issues/5068.

**Comment by wszymanski** [comment]
Sort indicator will be shown by default from now.

**Comment by wszymanski** [comment]
Sorting is passive from now. Any operation that affects sorted column data will turn off sorting ie. inserting a row, deleting a row, editing a cell, trimming row, moving row etc.

File: src/plugins/multiColumnSorting/domHelper.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+/**
+ * Helper for the column sorting plugin. Manages the added and removed classes to DOM elements basing on state of sorting.
+ *
+ * @class DomHelper
+ * @plugin MultiColumnSorting
+ */
+export class DomHelper {
+  constructor(columnStatesManager) {
+    this.columnStatesManager = columnStatesManager;
```
Comment (swistach) [inline src/plugins/multiColumnSorting/domHelper.js]: Could you add description what type it is?

File: src/plugins/multiColumnSorting/multiColumnSorting.js [hunk 2]
```diff
@@ -0,0 +1,804 @@
...
+   * Get if sort configs are valid.
+   *
+   * @private
+   * @param {Array} sortConfig Sort configuration for all sorted columns. Objects contain `column` and `sortOrder` properties.
+   * @returns {Boolean}
+   */
+  areValidSortConfigs(sortConfigs) {
+    const sortedColumns = sortConfigs.map(({ column }) => column);
+    const numberOfColumns = this.hot.countCols();
+
```
Comment (swistach) [inline src/plugins/multiColumnSorting/multiColumnSorting.js]: Is this empty line necessary?

File: src/plugins/multiColumnSorting/multiColumnSorting.js [hunk 3]
```diff
@@ -0,0 +1,804 @@
...
+      const numberOfColumns = this.hot.countCols();
+
+      rangeEach(numberOfColumns, visualColumnIndex => this.setMergedPluginSettings(visualColumnIndex));
+    }
+
+    const cellMeta = this.hot.getCellMeta(0, column);
+
+    this.blockPluginTranslation = false;
+
+    const cellMetaCopy = Object.create(cellMeta);
```
Comment (swistach) [inline src/plugins/multiColumnSorting/multiColumnSorting.js]: Maybe `const`s could be defined one by one?

File: src/plugins/multiColumnSorting/multiColumnSorting.js [hunk 4]
```diff
@@ -0,0 +1,804 @@
...
+  /**
+   * Callback for the `afterCreateRow` hook.
+   *
+   * @private
+   * @param {Number} index Visual index of the created row.
+   * @param {Number} amount Amount of created rows.
+   */
+  onAfterCreateRow(index, amount) {
+    this.rowsMapper.shiftItems(index, amount);
+
```
Comment (swistach) [inline src/plugins/multiColumnSorting/multiColumnSorting.js]: Is this empty line necessary?

File: src/plugins/multiColumnSorting/multiColumnSorting.js [hunk 5]
```diff
@@ -0,0 +1,804 @@
...
+   */
+  wasClickableHeaderClicked(event, column) {
+    const pluginSettingsForColumn = this.getFirstCellSettings(column).multiColumnSorting;
+    const headerActionEnabled = pluginSettingsForColumn.headerAction;
+
+    if (headerActionEnabled && event.realTarget.nodeName === 'SPAN') {
+      return true;
+    }
+
+    return false;
```
Comment (swistach) [inline src/plugins/multiColumnSorting/multiColumnSorting.js]: You can just return: `return headerActionEnabled && event.realTarget.nodeName === 'SPAN';`

File: src/plugins/multiColumnSorting/multiColumnSorting.js [hunk 6]
```diff
@@ -0,0 +1,804 @@
...
+    // Click below the level of column headers
+    if (coords.row >= 0) {
+      return;
+    }
+
+    if (this.wasClickableHeaderClicked(event, coords.col)) {
+      if (isPressedCtrlKey()) {
+        this.hot.deselectCell();
+        this.hot.selectColumns(coords.col);
+
```
Comment (swistach) [inline src/plugins/multiColumnSorting/multiColumnSorting.js]: Is this empty line necessary?

File: src/plugins/multiColumnSorting/test/multiColumnSorting.e2e.js [hunk 7]
```diff
@@ -0,0 +1,3180 @@
+describe('MultiColumnSorting', () => {
+  const id = 'testContainer';
+
+  beforeEach(function() {
+    this.$container = $(`<div id="${id}" style="overflow: auto; width: 300px; height: 200px;"></div>`).appendTo('body');
+
+    this.sortByClickOnColumnHeader = (columnIndex) => {
+      const hot = this.$container.data('handsontable');
+      const $columnHeader = $(hot.view.wt.wtTable.getColumnHeader(columnIndex));
+      const $spanInsideHeader = $columnHeader.find('.columnSorting');
```
Comment (swistach) [inline src/plugins/multiColumnSorting/test/multiColumnSorting.e2e.js]: What if `$columnHeader` would be `undefined`?

File: src/plugins/multiColumnSorting/test/multiColumnSorting.e2e.js [hunk 8]
```diff
@@ -0,0 +1,3180 @@
...
+
+      $spanInsideHeader.simulate('mousedown');
+      $spanInsideHeader.simulate('mouseup');
```
Comment (swistach) [inline src/plugins/multiColumnSorting/test/multiColumnSorting.e2e.js]: Could you add `$spanInsideHeader.simulate('click');`?

File: src/plugins/multiColumnSorting/test/multiColumnSorting.e2e.js [hunk 9]
```diff
@@ -0,0 +1,3180 @@
...
+      'Chuck Jackson',
+      'Rob Norris'
+    ]);
+
+    if (spec().$container) {
+      destroy();
+      spec().$container.remove();
+    }
+
+    hot = handsontable({
```
Comment (swistach) [inline src/plugins/multiColumnSorting/test/multiColumnSorting.e2e.js]: I'm not sure this should be `window.hot`.

File: src/plugins/multiColumnSorting/test/multiColumnSorting.e2e.js [hunk 10]
```diff
@@ -0,0 +1,3180 @@
...
```
Comment (swistach) [inline src/plugins/multiColumnSorting/test/multiColumnSorting.e2e.js]: As a few lines above.

File: src/plugins/multiColumnSorting/test/multiColumnSorting.e2e.js [hunk 11]
```diff
@@ -0,0 +1,3180 @@
...
+        'BMW1',
+        'Audi1'
+      ]);
+
+      if (spec().$container) {
+        destroy();
+        spec().$container.remove();
+      }
+
+      hot = handsontable({
```
Comment (swistach) [inline src/plugins/multiColumnSorting/test/multiColumnSorting.e2e.js]: As a few lines above.

**Comment by wszymanski** [comment]
Extra fix for rendering too wide header when `colHeaders` is set to `false`: https://github.com/handsontable/handsontable/commit/953c624d2325e20cbbd2d4d59d1305445d1d8ecd.

Use case found by @swistach.

sep-17-2018 12-34-55

**Comment by wszymanski** [comment]
New `setSortConfigs` method was implemented.

**Comment by wszymanski** [comment]
Extra warn about sort config validation was added.

**Pull request merged**